test/core/event_engine/test_suite/tools/echo_client.cc

```diff
@@ -0,0 +1,143 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#include <grpc/support/port_platform.h>
+
+// The echo client wraps an EventEngine::Connect and EventEngine::Endpoint
+// implementations, allowing third-party TCP listeners to interact with your
+// EventEngine client. Example usage:
+//
+//    # in one shell
+//    choco install nmap
+//    ncat -klp 32000
+//    # wait for a connection, than send data (e.g., keyboard input)
+//
+//    # in a separate shell
+//    bazel run
+//    //test/core/event_engine/test_suite/tools:my_event_engine_echo_client
+
+#include "absl/flags/flag.h"
+#include "absl/flags/parse.h"
+#include "absl/status/statusor.h"
+#include "absl/strings/str_format.h"
+
+#include <grpc/event_engine/event_engine.h>
+#include <grpc/event_engine/memory_allocator.h>
+#include <grpc/event_engine/slice_buffer.h>
+#include <grpc/grpc.h>
+#include <grpc/support/log.h>
+
+#include "src/core/lib/config/core_configuration.h"
+#include "src/core/lib/event_engine/channel_args_endpoint_config.h"
+#include "src/core/lib/event_engine/default_event_engine.h"
+#include "src/core/lib/event_engine/tcp_socket_utils.h"
+#include "src/core/lib/gprpp/notification.h"
+#include "src/core/lib/resource_quota/memory_quota.h"
+#include "src/core/lib/resource_quota/resource_quota.h"
+#include "test/core/event_engine/test_utils.h"
+
+extern absl::AnyInvocable<
+    std::unique_ptr<grpc_event_engine::experimental::EventEngine>(void)>
+CustomEventEngineFactory();
+
+ABSL_FLAG(std::string, target, "ipv4:127.0.0.1:50051", "Target string");
+
+namespace {
+using namespace std::chrono_literals;
+
+using ::grpc_event_engine::experimental::ChannelArgsEndpointConfig;
+using ::grpc_event_engine::experimental::EventEngine;
+using ::grpc_event_engine::experimental::GetDefaultEventEngine;
+using ::grpc_event_engine::experimental::Slice;
+using ::grpc_event_engine::experimental::SliceBuffer;
+using ::grpc_event_engine::experimental::URIToResolvedAddress;
+
+void SendMessage(EventEngine::Endpoint* endpoint, int message_id) {
+  SliceBuffer buf;
+  buf.Append(Slice::FromCopiedString(
+      absl::StrFormat("Waiting for message %d ... \n", message_id)));
+  grpc_core::Notification write_done;
+  endpoint->Write(
+      [&](absl::Status status) {
+        GPR_ASSERT(status.ok());
+        write_done.Notify();
+      },
+      &buf, nullptr);
+  write_done.WaitForNotification();
+}
+
+void ReceiveAndEchoMessage(EventEngine::Endpoint* endpoint, int message_id) {
+  SliceBuffer buf;
+  grpc_core::Notification read_done;
+  endpoint->Read(
+      [&](absl::Status status) {
+        if (!status.ok()) {
+          gpr_log(GPR_ERROR, "Error reading from endpoint: %s",
+                  status.ToString().c_str());
+          exit(1);
+        }
+        Slice received = buf.TakeFirst();
```

Comment:
Its not always guaranteed that the first slice read will contain the whole message appended in line 67. The ReceiveAndEchoMessage function should perhaps take an expected number of bytes to read as an argument and recursively call endpoint->Read until the expected number of bytes have been read

Reply:
Understood. I'll do a fix before we release it in a separate PR, I just pulled the first slice here for testing purposes.